zerver/lib/test_classes.py

```diff
@@ -1117,3 +1123,193 @@ def setUp(self) -> None:
+
+        self.login_user(user_profile)
+        result = self.do_email_change(user_profile, action_key)
+        self.assert_in_success_response(["This confirms that the email address for your Zulip"],
+                                        result)
+
+        self.verify_user_email_changed(action_key, user_profile.realm,
+                                       email_address_visibility_public=True)
+
+    def test_email_change_login_through_different_account(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+        old_email = user_profile.delivery_email
+
+        action_key = self.create_email_change_confirmation_key(user_profile, new_email)
+
+        different_profile = self.example_user('cordelia')
+        result = self.do_email_change(different_profile, action_key)
+        self.assertEqual(result.status_code, 400)
+        self.assert_in_response("Incorrect login. Login with your old email to change your email address",
+                                result)
+
+        with self.assertRaises(UserProfile.DoesNotExist):
+            get_user_by_delivery_email(new_email, user_profile.realm)
+        get_user_by_delivery_email(old_email, user_profile.realm)
+
+    def test_email_change_invalid_key(self) -> None:
+        user_profile = self.example_user('hamlet')
+
+        result = self.do_email_change(user_profile, 'invalid_action_key')
+        self.assert_in_success_response(["Whoops. The confirmation link is malformed."],
+                                        result)
+
+    def test_email_change_expired_key(self) -> None:
+        user_profile = self.example_user('hamlet')
+        new_email = 'hamlet-new@zulip.com'
+
+        date_sent = timezone_now() - datetime.timedelta(days=2)
+        action_key = self.create_email_change_confirmation_key(user_profile, new_email,
+                                                               date_sent=date_sent)
+
+        result = self.do_email_change(user_profile, action_key)
+        self.assert_in_success_response(["The confirmation link has expired or been deactivated."], result)
+
+    def test_email_change_unregistered_user_login(self) -> None:
+        '''
+        When trying to login with new email, the user shouldn't be taken to
+        registration page social and remote auth flows and email change should happen.
+        '''
+
+        # UserProfile is created here(not stored to db) as `do_change_email`
+        # expects a UserProfile
+        unregistered_user_profile = create_user_profile(get_realm('zulip'), email='new_email@zulip.com',
+                                                        password=None, active=False, bot_type=None,
+                                                        full_name='Nonexistent User', bot_owner=None,
+                                                        is_mirror_dummy=False, tos_version=None,
+                                                        timezone=None)
```

Comment:
would we be able to get rid of this hackiness if `do_change_email` took an email address (and realm?) rather than a userprofile?

Reply:
Ahh, yes, that sounds like a great idea. email and action key would be sufficient remote sso and email+password. We'd need name and subdomain for social auth.
I think sending those two as **kwargs would be a good idea as we don't need them in two of the auth flows. Does it make sense to you?

Reply:
Okay, changed to just send email and action key. This let us not get rid of the step of getting `email` from `user_profile` but since `name` was needed in social auth flow tests, made a `try` `except` block like this
```
    def do_email_change(self, email: str, action_key: str) -> HttpResponse:
        try:
            user_profile = get_user_by_delivery_email(email, get_realm("zulip"))
            name = user_profile.full_name
        except UserProfile.DoesNotExist:
            name = "Nonexistent User"
```